src/prop/proof_cnf_stream.cpp

```diff
@@ -0,0 +1,1265 @@
+      d_proof.addStep(clauseNode, PfRule::NOT_XOR_ELIM2, {node.notNode()}, {});
+      Trace("cnf")
+          << "ProofCnfStream::convertAndAssertXor: NOT_XOR_ELIM2 added "
+          << clauseNode << "\n";
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
+      if (Trace.isOn("cnf") && normClauseNode != clauseNode)
+      {
+        Trace("cnf")
+            << "ProofCnfStream::convertAndAssertXor: steps to normalized "
+            << normClauseNode << "\n";
+      }
+      MinisatSatSolver* minisat =
+          static_cast<MinisatSatSolver*>(d_cnfStream.d_satSolver);
+      minisat->getProofManager()->registerSatAssumptions({normClauseNode});
+    }
+    // Construct the clause ~q v p
+    SatClause clause2(2);
+    clause2[0] = p;
+    clause2[1] = ~q;
+    added = d_cnfStream.assertClause(node.negate(), clause2);
+    if (added)
+    {
+      Node clauseNode = nm->mkNode(kind::OR, node[0], node[1].notNode());
+      d_proof.addStep(clauseNode, PfRule::NOT_XOR_ELIM1, {node.notNode()}, {});
+      Trace("cnf")
+          << "ProofCnfStream::convertAndAssertXor: NOT_XOR_ELIM1 added "
+          << clauseNode << "\n";
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
+      if (Trace.isOn("cnf") && normClauseNode != clauseNode)
+      {
+        Trace("cnf")
+            << "ProofCnfStream::convertAndAssertXor: steps to normalized "
+            << normClauseNode << "\n";
+      }
+      MinisatSatSolver* minisat =
+          static_cast<MinisatSatSolver*>(d_cnfStream.d_satSolver);
+      minisat->getProofManager()->registerSatAssumptions({normClauseNode});
+    }
+  }
+}
+
+void ProofCnfStream::convertAndAssertIff(TNode node, bool negated)
+{
+  Trace("cnf") << "ProofCnfStream::convertAndAssertIff(" << node
+               << ", negated = " << (negated ? "true" : "false") << ")\n";
+  if (!negated)
+  {
+    // p <=> q
+    Trace("cnf") << push;
+    SatLiteral p = toCNF(node[0], false);
+    SatLiteral q = toCNF(node[1], false);
+    Trace("cnf") << pop;
+    bool added;
+    NodeManager* nm = NodeManager::currentNM();
+    // Construct the clauses ~p v q
+    SatClause clause1(2);
+    clause1[0] = ~p;
+    clause1[1] = q;
+    added = d_cnfStream.assertClause(node, clause1);
+    if (added)
+    {
+      Node clauseNode = nm->mkNode(kind::OR, node[0].notNode(), node[1]);
+      d_proof.addStep(clauseNode, PfRule::EQUIV_ELIM1, {node}, {});
+      Trace("cnf") << "ProofCnfStream::convertAndAssertIff: EQUIV_ELIM1 added "
+                   << clauseNode << "\n";
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
+      if (Trace.isOn("cnf") && normClauseNode != clauseNode)
+      {
+        Trace("cnf")
+            << "ProofCnfStream::convertAndAssertIff: steps to normalized "
+            << normClauseNode << "\n";
+      }
+      MinisatSatSolver* minisat =
+          static_cast<MinisatSatSolver*>(d_cnfStream.d_satSolver);
```

Comment:
Should you just do a single static_cast in the constructor?  Or better, have the constructor of this class take `MinisatSatSolver *`?

Reply:
I will actually make this class take the SatProofManager, that's the real dependency here.